spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java

```diff
@@ -560,42 +700,56 @@ protected AttributeAccessor getErrorMessageAttributes(Message<?> message) {
 	}
 
 	private void rethrow(Throwable t, String description) {
+		throw wrapExceptionIfNecessary(t, description);
+	}
+
+	private RuntimeException wrapExceptionIfNecessary(Throwable t, String description) {
 		if (t instanceof RuntimeException) {
-			throw (RuntimeException) t;
+			return (RuntimeException) t;
+		}
+		else {
+			return new MessagingException(description, t);
 		}
-		throw new MessagingException(description, t);
 	}
 
-	private void registerReplyMessageCorrelator() {
-		synchronized (this.replyMessageCorrelatorMonitor) {
-			if (this.replyMessageCorrelator != null) {
-				return;
-			}
-			AbstractEndpoint correlator = null;
-			BridgeHandler handler = new BridgeHandler();
-			if (this.getBeanFactory() != null) {
-				handler.setBeanFactory(this.getBeanFactory());
-			}
-			handler.afterPropertiesSet();
-			MessageChannel replyChannel = getReplyChannel();
-			if (replyChannel instanceof SubscribableChannel) {
-				correlator = new EventDrivenConsumer((SubscribableChannel) replyChannel, handler);
-			}
-			else if (replyChannel instanceof PollableChannel) {
-				PollingConsumer endpoint = new PollingConsumer((PollableChannel) replyChannel, handler);
-				endpoint.setBeanFactory(this.getBeanFactory());
-				endpoint.setReceiveTimeout(this.replyTimeout);
-				endpoint.afterPropertiesSet();
-				correlator = endpoint;
-			}
-			else {
-				throw new MessagingException("Unsupported 'replyChannel' type [" + replyChannel.getClass() + "]."
-						+ "SubscribableChannel or PollableChannel type are supported.");
-			}
-			if (this.isRunning()) {
-				correlator.start();
+	protected void registerReplyMessageCorrelatorIfNecessary() {
+		MessageChannel replyChannel = getReplyChannel();
+		if (replyChannel != null && this.replyMessageCorrelator == null) {
+			synchronized (this.replyMessageCorrelatorMonitor) {
```

Comment:
Review synchronized use:
when some code is synchronized, make sure is only doing local state mutation, then do call  user provided code outside, e.g. 
```boolean shouldStart = false;
    synchronized { //.. 
          shouldStart = isRunning();
     }
     if(shouldStart) {
        correlator.start();
     }
```